moped-toolbox/backfill_added_by/run.py

```diff
@@ -0,0 +1,184 @@
+        "moped_project"
+    ]
+    activities = make_hasura_request(
+        query=ACTIVITY_LOG_LOOKUP_QUERY, env=env, variables=None
+    )["moped_activity_log"]
+    users = make_hasura_request(query=MOPED_USERS_QUERY, env=env, variables=None)[
+        "moped_users"
+    ]
+
+    # we will attribute AMD projects to ivonne if no user info is available
+    amd_user_id = next(user for user in users if user["first_name"].lower() == "ivonne")
+    if not amd_user_id:
+        raise Exception("Unable to find default AMD user")
+
+    ready = []
+    unable_to_process = []
+
+    for proj in projects:
+        proj_id = proj["project_id"]
+        logs = [l for l in activities if l["record_project_id"] == proj["project_id"]]
+        if not logs:
+            # some projects are missing logs - presumably to due SQS/event handler failures
+            if proj["date_added"].startswith("2022-02-10"):
+                # this is an AMD project migrated on 2 Feb 2022
+                ready.append({"project_id": proj_id, "added_by": amd_user_id})
+                continue
+            else:
+                unable_to_process.append(
+                    {"project_id": proj_id, "reason": "no 'insert' event log found"}
+                )
+                continue
+        # ignore extra logs (there may be dupes due to SQS/event handler bug )
+        log = logs[0]
+        user = log["moped_user"]
+
+        if not user:
+            if proj["date_added"].startswith("2022-02-10"):
+                # this is an AMD project migrated on 2 Feb 2022 - it was created without a session token so no user ID was logged
+                ready.append({"project_id": proj_id, "added_by": amd_user_id})
+                continue
+            elif log["updated_by"]:
+                # we logged a cognito user ID but it is no longer associated with a user in the DB
+                unable_to_process.append(
+                    {"project_id": proj_id, "reason": "unknown cognito user ID"}
+                )
+                continue
+            else:
+                unable_to_process.append(
+                    {"project_id": proj_id, "reason": "no user data"}
+                )
+                continue
+        # this project is ready to go
+        user_id = user["user_id"]
+        ready.append({"project_id": proj_id, "added_by": user_id})
+
+    counts = {
+        "updated": 0,
+        "todo": len(ready),
+        "unable_to_process": len(unable_to_process),
+    }
+
+    print(counts)
+
+    for proj in ready:
+        make_hasura_request(
+            query=UPDATE_PROJECT_ADDED_BY_MUTATION,
+            variables={"project_id": proj["project_id"], "added_by": proj["added_by"]},
+            env=env,
+        )
+        counts["updated"] += 1
+        counts["todo"] -= 1
+        print(counts)
+
+    print("Writing inoperable projects to 'log.json'")
+
+    with open("log.json", "w") as fout:
+        json.dump(unable_to_process, fout)
```

Comment:
This is probably not needed as I imagine you have a pretty good idea of what will be in the `log.json` file for our production instance. I like the pattern here of using a JSON log format, and the only thing I can think to even suggest is to maybe use something like `json.dumps(object, indent=2)` to increase its readability.

Reply:
+1 this a TIL for me